### ISSUE TITLE
Is launchAff too dangerous?

### ISSUE THREAD

**Issue by hdgarrood** [slamdata/purescript-aff#32]
Currently, `launchAff` throws away all errors. This might not be so bad in a browser, as the errors might appear in the console anyway? It's particularly bad on node, though, as there's no indication at all that an error has occurred.

I am aware that this is intended and documented behaviour, but I think it's still very easy to be tripped up by this. I think there's a case to be made that it's too dangerous to be exposed as is, especially since if users really want that functionality, they can easily implement it themselves in terms of `runAff`.

What do you think about, say, `launchAff = runAff throwException (const (pure unit))` instead?

**Comment by beckyconning** [comment]
:+1:

**Comment by jdegoes** [comment]
:+1: 

Another option would be using the global error handler, although I don't know if that's portable to node. But I think I like just throwing the exception better.

**Comment by hdgarrood** [comment]
I just tried this and it's a little harder than I thought to make it typecheck, because `runAff` forces you to have the same row of effects in the `Eff` you get out. I want to write this:

```purescript
  -- | Converts the asynchronous computation into a synchronous one. All values
  -- | are ignored, and if the computation produces an error, it is thrown.
  launchAff :: forall e a. Aff e a -> Eff (err :: EXCEPTION | e) Unit
  launchAff = runAff throwException (const (pure unit)) <<< liftEx
    where
    liftEx :: forall e a. Aff e a -> Aff (err :: EXCEPTION | e) a
    liftEx x = x
```

But `liftEx` doesn't typecheck. Any suggestions?

**Comment by paf31** [comment]
`unsafeInterleaveEff`?

**Comment by garyb** [comment]
If you defined it as 
``` purescript
launchAff :: forall e a. Aff (err :: EXCEPTION | e) a -> Eff (err :: EXCEPTION | e) Unit
```

That would work too right? You don't _have_ to use the effect just because it's there.

**Comment by hdgarrood** [comment]
Yes, but I think that would prevent you from using a closed row of effects for an `Aff` action unless that row contained `err :: EXCEPTION`, which doesn't seem right, as it's an implementation detail, and since `Aff` already includes its own error handling.

I guess `unsafeInterleaveEff` works. Is there no "above-board" way of doing this?

**Comment by beckyconning** [comment]
Are we dead set on keeping launchAff?

**Comment by garyb** [comment]
For the closed row case I guess you could go via `runAff`?

**Comment by hdgarrood** [comment]
> Are we dead set on keeping launchAff?

I think it's worth having the version of `launchAff` that throws exceptions - it seems like a very useful thing. I expect most uses of `Aff` will have a `runAff` or `launchAff` to kick the whole thing off. Also, as I'm discovering, it's not trivial to write the throwing version!

> For the closed row case I guess you could go via runAff?

You could, but I think it's an unnecessary leakage of implementation details into the API.

**Comment by hdgarrood** [comment]
Oh wait, could you? I think I might have misunderstood your comment.

**Comment by garyb** [comment]
No, probably not actually, it'd have the same issue. I guess the problem that brought us to this is uncaught stack overflows, however that's not an effect that we have any way of explicitly tracking. Wouldn't it be nice if we actually could do that? :wink:

**Comment by hdgarrood** [comment]
I think this can be a problem in other cases too. When this bit me, it wasn't stack overflows, it was me using the API of a node package incorrectly (node-irc).

**Comment by hdgarrood** [comment]
This was resolved by #33.
